Include browser performance values on $pageview

pauldambra opened the pull request:
## Changes

Captures browser performance measurements available when we send a pageview.

Browsers capture [page performance information](https://developer.mozilla.org/en-US/docs/Web/API/Performance) behind the read-only property `window.performance`

The value captured only consistently included timing information when the performance object was stringified. This adds 17 milliseconds to the pageview event generation.

Would need a plugin for processing the values to something more useful

(Also updates Cypress to latest version)

## Checklist
- [x] Tests for new code (if applicable)

macobo commented on cypress/integration/capture.spec.js:
```diff
@@ -194,6 +194,45 @@ describe('Event capture', () => {
+
+                expect(captures['event']).to.equal('$pageview')
+
+                expect(captures?.properties).to.have.property('performance')
```
Excess ? everywhere in the added tests - these shouldn't be here.

macobo commented:
We have 3 tiers of tests in posthog-js:

1. Unit tests - this verifies the behavior of the library in bite-sized chunks. Keep this coverage as high as possible, easiest to test corner cases with
2. Cypress tests - integrates with a real browser and is capable of testing timing, browser requests, etc. Super useful for testing high-level ordering and verifying some behavior, but we shouldn't aim for 100% coverage here
3. E2E testcafe tests - integrates with a real posthog instance sending data to it.

In this case you're testing most things at layer 2, some at layer 3.

Instead I suggest:
- Have proper unit test coverage at (1), testing behavior both the feature on/off
- If you know the actual structure of what the key will contain beyond the 3 keys, add a cypress test asserting it's structure
- Testcafe test isn't likely to be useful unless the code may throw in some browsers (e.g. is `window.performance` defined in ie8?)

macobo reviewed (commented):
I think the tests are a m

pauldambra commented:
> Have proper unit test coverage at (1), testing behavior both the feature on/off

added unit tests

> If you know the actual structure of what the key will contain beyond the 3 keys, add a cypress test asserting it's structure

the structure is pretty large, which is why I avoided adding too much to the test

<img width="405" alt="Screenshot 2021-12-22 at 10 57 57" src="https://user-images.githubusercontent.com/984817/147089546-8f7c29b0-7385-4b88-bac9-e40be1b5142a.png">

> Testcafe test isn't likely to be useful unless the code may throw in some browsers (e.g. is window.performance defined in ie8?)

Yeah, good prompt. It isn't available pre-IE9

I've made the code safer and removed the testcafe test.

macobo commented on src/posthog-core.js:
```diff
@@ -290,7 +300,16 @@ PostHogLib.prototype._loaded = function () {
+        const props = {}
+        if (this.get_config('capture_performance')) {
+            // TRICKY: without the JSON stringing and parsing less perf data is sent
+            props.performance = {
```
Problem: Posthog kind of sucks at querying nested values right now.

pauldambra replied:
Yep, I didn't think it was right to do processing in the SDK. So my intention is to pair this with a plugin. We'll keep the "raw" performance data in case it is useful in future but use it to add useful queryable fields on ingestion

macobo replied:
That sounds like the feature is pretty alpha-ish. Q: Should we make the config value indicate that somehow (e.g. underscore it)?

pauldambra replied:
Definitely alpha :) I'm crossing the river by feeling the stones.

macobo commented on src/__tests__/posthog-core.js:
```diff
@@ -677,24 +678,128 @@ describe('_loaded()', () => {
+            })
+
+            try {
+                given('config', () => ({
```
Tests should follow the structure setup -> action -> test

Within this block config does not belong in a try-catch. Given this config is also repeated for most of your tests, set this config at the describe level and only override it at the single disabled test.

macobo commented on src/__tests__/posthog-core.js:
```diff
@@ -677,24 +678,128 @@ describe('_loaded()', () => {
+                )
+                expect(performance.getEntriesByType).not.toHaveBeenCalled()
+            } finally {
+                window.performance = performance
```
Your tests have really messy setup code. We can clean this up really nicely by leveraging beforeEach/afterEach is for. Tests should only test what's relevant.

Suggestion:

```
beforeEach(() => {
  jest.spyOn(window.performance, 'getEntriesByType').mockImplementation(() => given.performanceEntries())
})
```

And use given in each of your tests.

In general having try/catch in your tests is a definite smell.

pauldambra replied:
have added an extra level of describe to move some of the setup and teardown out of the tests

we can't spy on performance `getEntriesByType` directly (because of a JSDOM issue https://github.com/jsdom/jsdom/issues/3309) so have followed the advice here: https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom

macobo replied:
You can simplify this further by doing this in your definition then:

```js
Object.defineProperty(window, 'performance', {
    writable: true,
    value: {
        getEntriesByType: jest.fn().mockImplementation(() => given.performanceEntries()),
    }
})
```

This would clean up the `'safely attempts to capture performance if a type of entry is not available in a browser'` test.

macobo commented on cypress/integration/capture.spec.js:
```diff
@@ -194,6 +196,45 @@ describe('Event capture', () => {
+            })
+        })
+
+        it('does not send performance timing with the pageview when disabled', () => {
```
I'd remove this test - this is sufficiently tested by unit tests.

macobo commented on cypress/integration/capture.spec.js:
```diff
@@ -194,6 +196,45 @@ describe('Event capture', () => {
+                expect(captures['event']).to.equal('$pageview')
+
+                expect(captures.properties).to.have.property('performance')
+                expect(captures.properties.performance).to.have.property('navigation')
```
Cypress tests run in chrome and we can make use of that fact here given that the contents of the performance keys are really opaque (when do we know it breaks?)

I think we can make this test more useful by asserting that each of the properties contains at least some keys that will be later used in analysis.

pauldambra replied:
have added some more detail to the tests and a comment on why I'm being slightly cautious

macobo commented on src/posthog-core.js:
```diff
@@ -290,7 +300,16 @@ PostHogLib.prototype._loaded = function () {
-        this.capture('$pageview', {}, { send_instantly: true })
+        const props = {}
+        if (this.get_config('capture_performance')) {
+            // TRICKY: without the JSON stringing and parsing less perf data is sent
```
This comment is now out of place

macobo commented on src/posthog-core.js:
```diff
@@ -106,6 +106,7 @@ const defaultConfig = () => ({
         const error = 'Bad HTTP status: ' + req.status + ' ' + req.statusText
         console.error(error)
     },
+    capture_performance: false,
```
Did you update ts types?

macobo commented on cypress/integration/capture.spec.js:
```diff
@@ -194,6 +196,45 @@ describe('Event capture', () => {
+            cy.get('@capture').should(({ request }) => {
+                const captures = getBase64EncodedPayload(request)
+
+                expect(captures['event']).to.equal('$pageview')
```
Note: This looks inconsistent if we're accessing some properties via [] and others via . - pick one.

macobo commented on src/__tests__/posthog-core.js:
```diff
@@ -677,24 +678,121 @@ describe('_loaded()', () => {
+            beforeEach(() => {
+                performance = { ...window.performance }
+
+                given('config', () => ({
```
You can put this outside of beforeEach.

macobo replied:
This does not need to be in a beforeEach. See also https://github.com/tatyshev/given2#more-examples

macobo commented on src/__tests__/posthog-core.js:
```diff
@@ -677,24 +678,121 @@ describe('_loaded()', () => {
-        }))
+    describe('capturing pageviews', () => {
+        beforeEach(() => {
+            given('config', () => ({
```
This does not need to be in a beforeEach. See also https://github.com/tatyshev/given2#more-examples

macobo commented on src/__tests__/posthog-core.js:
```diff
@@ -677,24 +678,121 @@ describe('_loaded()', () => {
+        beforeEach(() => {
+            given('config', () => ({
+                capture_pageview: false,
+                capture_performance: false, // the default
```
No need to set this at all if it's the default?

macobo commented on src/__tests__/posthog-core.js:
```diff
@@ -677,24 +678,121 @@ describe('_loaded()', () => {
-            capture_pageview: false,
-        }))
+        describe('when window performance is available', () => {
+            it('captures pageview', () => {
```
This test is in the wrong place. This should be one describe up - performance logic does not affect whether we capture a pageview or not.

macobo commented on cypress/integration/capture.spec.js:
```diff
@@ -194,6 +196,74 @@ describe('Event capture', () => {
+                expect(performance).to.have.property('paint')
+                expect(performance.paint).to.be.instanceof(Array)
+                if (performance.paint.length > 0) {
+                    // we can't guarantee we run early enough to capture paint results
```
This comment seems self-contradictory?

pauldambra replied:
Ah, it's to explain why the tests are in an if condition at all. I'll clarify that

macobo commented on cypress/integration/capture.spec.js:
```diff
@@ -194,6 +196,74 @@ describe('Event capture', () => {
+            })
+        })
+
+        it('does not send performance timing with the pageview when disabled', () => {
```
Is this a useful test? I don't think it is - we're asserting the same behavior in unit tests, browser context gives us nothing extra.

macobo reviewed (approved):
More test comments but other than those don't let yourself be blocked.

macobo commented on src/posthog-core.js:
```diff
@@ -620,6 +612,15 @@ PostHogLib.prototype.capture = addOptOutCheckPostHogLib(function (event_name, pr
+            paint: getPerformanceEntriesByType('paint'),
+            resource: getPerformanceEntriesByType('resource'),
+        }
+        data.properties.performance = { ...data.properties.performance, ...performance }
```
Suggestion: Given that this is an internal property that won't be usable by outside users, prefix this with a $

macobo commented on src/posthog-core.js:
```diff
@@ -620,6 +612,15 @@ PostHogLib.prototype.capture = addOptOutCheckPostHogLib(function (event_name, pr
         properties: this._calculate_event_properties(event_name, properties, start_timestamp),
     }
 
+    if (event_name === '$pageview' && this.get_config('_capture_performance')) {
```
In an ideal world, methods do as little as possible - mixing business logic with capturing here is messy.

There is a call to `_calculate_event_properties` just above - extend that.

macobo reviewed (commented):
We've now iterated away from the previous approach - hence needing to review this all over again. This PR is getting really messy to follow, which hurts us both - let me know when it's ready for a review.

pauldambra commented:
> We've now iterated away from the previous approach

Yep, was just about to message you. Realised we didn't capture synthetic pageviews with the previous mechanism

pauldambra commented:
> We've now iterated away from the previous approach

reverted to the approved state and will follow-up with changes
